subprojects/scheduler/src/main/java/org/gradle/internal/scheduler/DefaultScheduler.java

```diff
@@ -0,0 +1,587 @@
+                scheduleWork(graph, nodeExecutor);
+            }
+            if (cancelled) {
+                break;
+            }
+            expectAvailableWorkers = handleEvents(graph, continueOnFailure, executedNodes, failures);
+        }
+    }
+
+    private void scheduleWork(final Graph graph, final NodeExecutor nodeExecutor) {
+        Queue<Node> rootNodes = graph.queueRootNodes();
+        System.out.printf(">> Scheduling root nodes: %s%n", rootNodes);
+
+        if (cancellationToken.isCancellationRequested()) {
+            cancelExecution(graph, false);
+            return;
+        }
+
+        if (rootNodes.isEmpty()) {
+            return;
+        }
+
+        while (true) {
+            // Find a root node to allocate
+            Node nodeToRun = rootNodes.poll();
+            if (nodeToRun == null) {
+                break;
+            }
+            if (runningNodes.contains(nodeToRun)) {
+                continue;
+            }
+
+            // TODO These could be handled outside of the coordination-service lock
+            if (!nodeToRun.getState().isExecutable()) {
+                enqueue(new NodeFinishedEvent(nodeToRun));
+                continue;
+            }
+
+            // Check if there is a node conflicting with the current one
+            Node conflictingNode = concurrentNodeExecutionCoordinator.findConflictingNode(graph, nodeToRun, runningNodes);
+            if (conflictingNode != null) {
+                // Make sure we don't retry this node until the conflicting node is finished
+                System.out.printf(">> Found conflict between %s and %s%n", nodeToRun, conflictingNode);
+                graph.addEdge(new Edge(conflictingNode, MUST_NOT_RUN_WITH, nodeToRun));
+                continue;
+            }
+
+            // Allocate a worker
+            TaskExecutorWorker worker = availableWorkers.poll();
+            if (worker == null) {
+                break;
+            }
+
+            // Run the node
+            System.out.printf(">> Trying to run %s on %s...%n", nodeToRun, Thread.currentThread().getName());
+            // TODO Don't do this for revived nodes that have already been prepared once?
+            prepareToRunNode(nodeToRun, graph, rootNodes);
+            runningNodes.add(nodeToRun);
+            ResourceLock nodeLock = concurrentNodeExecutionCoordinator.findLockFor(nodeToRun);
+            worker.run(new NodeExecution(nodeToRun, nodeExecutor, nodeLock));
+        }
+    }
+
+    private boolean handleEvents(Graph graph, boolean continueOnFailure, ImmutableList.Builder<Node> executedNodes, ImmutableList.Builder<Throwable> failures) {
+        try {
+            // Wait for at least one event to arrive
+            Event event = eventQueue.take();
+            eventsBeingProcessed.add(event);
+            // Make sure we take all available events
+            eventQueue.drainTo(eventsBeingProcessed, EVENTS_TO_PROCESS_AT_ONCE - 1);
```

Comment:
Why don't we process the events one by one? For performance reasons? Or are there some other reasons?

Reply:
I'm also wondering what the right amount of stuff to do both in scheduling new work and in processing events. There has to be a balance, but not sure how to keep that up. We now try to schedule work eagerly, doing it as much as possible, so it made sense to try to do the same with processing events.